### DISCUSSION TITLE
multiple nextjs on same domain

### DISCUSSION THREAD

**Discussion by always-maap** [vercel/next.js#34616] [Help] [answered]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP PREEMPT Sun Sep 19 21:31:53 UTC 2021
    Binaries:
      Node: 16.12.0
      npm: 8.1.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.0.11-canary.15
      react: 17.0.2
      react-dom: 17.0.2

### Describe the Bug

We have two nextjs applications on the same domain. one with mobile env and one with desktop env to save bytes. at the built time the ternary components remove unused code. we are using nginx to separate by user-agent. also the page is using getStaticProps.
```jsx
process.env.PLATFORM === 'mobile' ? <MobileCmp /> : <DesktopCmp />
```
![2022-02-20-232057_1920x1080_scrot](https://user-images.githubusercontent.com/58308349/154861481-66d1b105-e45c-469d-ab80-6bd488777c2f.png)

we are getting 404 for a couple of pages in desktop but on mobile everything works. or vice versa sometimes.
I found [a blog](https://arihantverma.com/posts/2021/07/18/workaround-nextjs-data-urls-same-domain-multiple-apps/) describing the same issue but after all his solution is a workaround and I wanna make sure there isn't any official way to solve it by the nextjs team

### Expected Behavior

I don't really. maybe it is something expected by nextjs team but for us as users, more clear way to solve it in doc or error message would be great. or maybe expose `data fetch request path URL prefix`.

### To Reproduce

NEXT_PUBLIC_PLATFORM=mobile npm run zoomit:serve -- port 4201
NEXT_PUBLIC_PLATFORM=mobile npm run zoomit:serve -- port 4202

separate by user agent in nginx

**Comment by balazsorban44** [comment]
Not sure I understand correctly, but are you maybe referring to Multi Zones?


https://nextjs.org/docs/advanced-features/multi-zones

If you want to split your users based on the incoming request, a Middleware with a rewrite is possibly the easiest way though.

https://nextjs.org/docs/middleware

**Reply by always-maap** [reply]
Nah, multi-zones are for another app for serving all other pages. but the problem is with the collusion of assets with the same pages as described [here](https://arihantverma.com/posts/2021/07/18/workaround-nextjs-data-urls-same-domain-multiple-apps/).

**Reply by balazsorban44** [reply]
Then I would recommend Middleware as mentioned above. You don't need two Next.js instances, just one that does a rewrite based on the incoming request.

You could also do that easily with nginx on top if you wanted to.

**Reply by always-maap** [reply]
I honestly read the middleware doc and all examples a couple of times but still, I don't understand what you mean by "rewrite based on the incoming request".
Can you explain more? Did I provide enough information about the problem we are facing?
basically, two apps run on the same domain and are separated by user-agent. the routes are the same. 
`localhost:4201/mobile` -> desktop 
`localhost:4202/mobile` -> mobile
and nginx reverse both to `https://domain.ir/mobile`. there is one codebase with progress.env.PLATFORM and a bunch of ternaries.

Thanks <333

**Reply by always-maap** [reply]
I think a couple of issues were closed but I couldn't find the best patterns for this problem.

related:
https://github.com/vercel/next.js/issues/257
https://github.com/vercel/next.js/issues/5602

thanks <333

@Timer @leerob

**Comment by balazsorban44** [comment, marked as answer]
After some discussion, we figured a solution together with @always-maap that would not require multiple apps in the first place, and thus also halves the build time:

https://github.com/balazsorban44/nextjs-34616

**Reply by always-maap** [reply]
Not only that but as you mentioned `outputStandalone: true` will eliminate a couple of steps in dockerizing and save so much time too. You are ridiculously talented. It takes you like 10min to reproduce, create a repo and deploy the working solution. much appreciated <333

**Comment by swati61** [comment]
Is there  any way to create a common Header and access that into another module in nextjs 12?
